src/memo/memo.mli

```diff
@@ -170,14 +170,34 @@ end
     cached. *)
 exception Non_reproducible of exn
 
-(** Notify the memoization system that the build system has restarted. This
-    removes the values that depend on the [current_run] from the memoization
-    cache, and cancels all pending computations. *)
-val reset : unit -> unit
+(** [Invalidation] describes a set of nodes to be invalidated between
+    memoization runs. These sets can be combined into larger sets to then be
+    passed to [reset]. *)
+module Invalidation : sig
+  type t
+
+  val empty : t
+
+  val combine : t -> t -> t
+
+  val is_empty : t -> bool
 
-(** Notify the memoization system that the build system has restarted but do not
-    clear the memoization cache. *)
-val restart_current_run : unit -> unit
+  (** Indicates that memoization tables should be cleared. We use it if
+      incremental mode is not enabled.
+
+      Bug: this is not sufficient to guarantee full recomputation because it
+      does not invalidate individual nodes, only tables, so if you hold on to a
+      node (via a [lazy_] or [cell] call), then that's going to keep its
+      potentially stale value. *)
+  val clear_caches : t
+end
+
+(** Notify the memoization system that the build system has restarted. This
+    removes the values specified by [Invalidation.t] from the memoization cache,
+    advances the current run, and cancels all pending computations. (*
+    CR-someday aalekseyev: In what sense does it cancel pending computations? I
+    don't think it does anything for that. *) *)
```

Comment:
as discussed offline, the comment about cancelling pending computations is either stale or was wrong from the start.